cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1125 @@
+/**
+ * \brief
+ * Source-OrderBySink
+ * In this example, the data enters through the source node
+ * and the data is ordered in the sink node. The order can be
+ * ASCENDING or DESCENDING and it is configurable. The output
+ * is obtained as a table from the sink node.
+ * \param exec_context : execution context
+ * \return arrow::Status
+ */
+arrow::Status SourceOrderBySinkExample(cp::ExecContext& exec_context) {
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+                        cp::ExecPlan::Make(&exec_context));
+
+  ARROW_ASSIGN_OR_RAISE(auto basic_data, MakeSortTestBasicBatches());
+
+  std::cout << "basic data created" << std::endl;
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+  auto source_node_options = cp::SourceNodeOptions{basic_data.schema, basic_data.gen()};
+  ARROW_ASSIGN_OR_RAISE(cp::ExecNode * source,
+                        cp::MakeExecNode("source", plan.get(), {}, source_node_options));
+
+  ARROW_ASSIGN_OR_RAISE(
+      std::ignore,
+      cp::MakeExecNode("order_by_sink", plan.get(), {source},
+                       cp::OrderBySinkNodeOptions{
+                           cp::SortOptions{{cp::SortKey{"a", cp::SortOrder::Descending}}},
+                           &sink_gen}));
+
+  // // // translate sink_gen (async) to sink_reader (sync)
+  std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+      basic_data.schema, std::move(sink_gen), exec_context.memory_pool());
+
+  // // // start the ExecPlan
+  ARROW_RETURN_NOT_OK(plan->StartProducing());
+
+  // // collect sink_reader into a Table
+  std::shared_ptr<arrow::Table> response_table;
+
+  ARROW_ASSIGN_OR_RAISE(response_table,
+                        arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+  std::cout << "Results : " << response_table->ToString() << std::endl;
+
+  return arrow::Status::OK();
+}
+
+// (Doc section: OrderBySink Example)
+
+// (Doc section: HashJoin Example)
+/**
+ * \brief
+ * Source-HashJoin-Sink
+ * This example shows how source node gets the data and how a self-join
+ * is applied on the data. The join options are configurable. The output
+ * is obtained as a table via the sink node.
+ * \param exec_context : execution context
+ * \return arrow::Status
+ */
+arrow::Status SourceHashJoinSinkExample(cp::ExecContext& exec_context) {
+  ARROW_ASSIGN_OR_RAISE(auto input, MakeGroupableBatches());
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+                        cp::ExecPlan::Make(&exec_context));
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+  cp::ExecNode* left_source;
+  cp::ExecNode* right_source;
+  for (auto source : {&left_source, &right_source}) {
+    ARROW_ASSIGN_OR_RAISE(*source,
+                          MakeExecNode("source", plan.get(), {},
+                                       cp::SourceNodeOptions{input.schema, input.gen()}));
+  }
+  // TODO: decide whether to keep the filters or remove
```

Comment:
Have you decided?

Reply:
Better to remove, we only need to show how a join works.